src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -21,49 +19,56 @@
         {
-            // Get the diagnostics that indicate a missing import.
-            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
-            if (diagnostics.IsEmpty)
+            var analysisResult = await AnalyzeAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return await AddMissingImportsAsync(analysisResult, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc/>
+        public async Task<Document?> AddMissingImportsAsync(AddMissingImportsAnalysisResult analysisResult, CancellationToken cancellationToken)
+        {
+            if (analysisResult.CanAddMissingImports)
             {
-                return false;
+                // Apply those fixes to the document.
+                var newDocument = await ApplyFixesAsync(analysisResult.Document, analysisResult.AddImportFixData, cancellationToken).ConfigureAwait(false);
+                return newDocument;
             }
 
-            // Find fixes for the diagnostic where there is only a single fix.
-            var usableFixes = await GetUnambiguousFixesAsync(document, diagnostics, cancellationToken).ConfigureAwait(false);
-            return !usableFixes.IsEmpty;
+            return null;
         }
 
-        public async Task<Project> AddMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        /// <inheritdoc/>
+        public async Task<AddMissingImportsAnalysisResult> AnalyzeAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
         {
             // Get the diagnostics that indicate a missing import.
             var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
             if (diagnostics.IsEmpty)
             {
-                return document.Project;
+                return new AddMissingImportsAnalysisResult(
+                    ImmutableArray<AddImportFixData>.Empty,
+                    document);
             }
 
             // Find fixes for the diagnostic where there is only a single fix.
             var unambiguousFixes = await GetUnambiguousFixesAsync(document, diagnostics, cancellationToken).ConfigureAwait(false);
 
             // We do not want to add project or framework references without the user's input, so filter those out.
             var usableFixes = unambiguousFixes.WhereAsArray(fixData => DoesNotAddReference(fixData, document.Project.Id));
-            if (usableFixes.IsEmpty)
-            {
-                return document.Project;
-            }
 
-            // Apply those fixes to the document.
-            var newDocument = await ApplyFixesAsync(document, usableFixes, cancellationToken).ConfigureAwait(false);
-            return newDocument.Project;
+            return new AddMissingImportsAnalysisResult(
+                usableFixes,
+                document);
```

Comment:
i don't understand why the result needs to hold the Doucment.  It's an input to the algorithm,a nd is not transformed.  So it doesn't need to be in the output.

Reply:
This pattern helps reduce errors. There's no way to validate that the input to the algorithm matched the input to the next call (in this case, using the AnalysisResult), and keeping it in the result object takes that possibility away. This strongly declares that the input must be the same when passing in the object rather than allowing it to split. It also makes for a cleaner internal API surface (my opinion).

Reply:
nothing else we have works this way.  If you do want to do things this way, then a better way would be to have the AddImportFixData hold onto this.